Whiskey/Functions/Resolve-WhiskeyTaskPathParameter.ps1

```diff
@@ -29,125 +244,235 @@ function Resolve-WhiskeyTaskPathParameter
+                $normalizedPath = Join-Path -Path $currentDir -ChildPath $result
             }
 
-        if( -not $result ) 
-        {
-            if( -not $ValidateAsPathAttribute.AllowNonexistent )
+            # Remove all the '..' and '.' path parts from the path.
+            if( -not [wildcardpattern]::ContainsWildcardCharacters($normalizedPath) )
             {
-                Stop-WhiskeyTask -TaskContext $TaskContext `
-                                 -Message ('{0}[{1}] "{2}" does not exist.' -f $CmdParameter.Name,$pathIdx,$Path)
-                return
+                $normalizedPath = [IO.Path]::GetFullPath($normalizedPath)
             }
-            $result = [IO.Path]::GetFullPath($Path)
 
-            if( [Management.Automation.WildCardPattern]::ContainsWildcardCharacters($result) )
+            if( (Test-Path -Path $normalizedPath) )
             {
-                return
+                $resolvedPaths = Get-Item -Path $normalizedPath -Force | Select-Object -ExpandProperty 'FullName'
             }
-        }
 
-            
-        if( -not $ValidateAsPathAttribute.AllowOutsideBuildRoot )
-        {
-            $fsCaseSensitive = -not (Test-Path -Path ($TaskContext.BuildRoot.FullName.ToUpperInvariant()))
-            $normalizedBuildRoot = $TaskContext.BuildRoot.FullName.TrimEnd([IO.Path]::DirectorySeparatorChar, [IO.Path]::AltDirectorySeparatorChar)
-            $normalizedBuildRoot = '{0}{1}' -f $normalizedBuildRoot,[IO.Path]::DirectorySeparatorChar
-
-            $comparer = [System.StringComparison]::OrdinalIgnoreCase
-            if( $fsCaseSensitive )
+            if( -not $resolvedPaths ) 
             {
-                $comparer = [System.StringComparison]::Ordinal
+                if( -not $AllowNonexistent )
+                {
+                    Stop-WhiskeyTask -TaskContext $TaskContext `
+                                    -Message ('{0}[{1}] "{2}" does not exist.' -f $PropertyName,$pathIdx,$Path)
+                    return
+                }
+
+                $resolvedPaths = $normalizedPath
+
+                # If it contains a wildcard, it didn't resolve to anything, so don't return it.
+                if( [wildcardpattern]::ContainsWildcardCharacters($resolvedPaths) )
+                {
+                    return
+                }
+
             }
             
-            $invalidPaths =
-                $result |
-                Where-Object { -not ( $_.StartsWith($normalizedBuildRoot, $comparer) ) }
-
-            if( $invalidPaths )
+            if( -not $AllowOutsideBuildRoot )
             {
-                Stop-WhiskeyTask -TaskContext $TaskContext `
-                                 -Message ('{0}[{1}] "{2}" is outside the build root "{3}".' -f $CmdParameter.Name,$pathIdx,$Path,$TaskContext.ConfigurationPath)
-                return
+                $fsCaseSensitive = -not (Test-Path -Path ($TaskContext.BuildRoot.FullName.ToUpperInvariant()))

```

Comment:
Feels like $normalizedBuildRoot assignments could go below the $compare if statement (so $fsCaseSensitive moves closer to single usage), or $fsCaseSensitive can move down for same effect.

Reply:
How can we be fooled by `C:\AWESOME\ME`? If we're on Windows, doesn't matter. If we're on Linux, path would be `/AWESOME/ME`. If I create an `/awesome/me/attackerfile`, `/awesome/me/` would not match `/AWESOME/ME/`, and the attacker is thwarted.